File: platforms/android/App/aj/src/main/java/applica/aj/runtime/javascriptcore/AJJavaScriptCoreRuntime.java
package applica.aj.runtime.javascriptcore;

import android.content.Context;
import android.util.Log;

import org.apache.commons.io.IOUtils;
import org.liquidplayer.webkit.javascriptcore.JSContext;
import org.liquidplayer.webkit.javascriptcore.JSException;
import org.liquidplayer.webkit.javascriptcore.JSFunction;
import org.liquidplayer.webkit.javascriptcore.JSObject;
import org.liquidplayer.webkit.javascriptcore.JSValue;

import java.io.InputStream;

import applica.aj.AJObject;
import applica.aj.Async;
import applica.aj.Semaphore;
import applica.aj.runtime.AJRuntime;
import applica.aj.runtime.Plugin;

/**
 * Created by bimbobruno on 10/03/16.
 */
public class AJJavaScriptCoreRuntime extends AJRuntime {

    public interface NativeGlobals {

        JSValue __requireInternal(String path);

        void async(JSValue action);

        void trigger(String type, JSValue data);

        void exec(String plugin, String fn, JSValue data, JSFunction callback);
    }

    public class NativeGlobalsImpl extends JSObject implements NativeGlobals {

        public NativeGlobalsImpl(JSContext jsContext) {
            super(jsContext, NativeGlobals.class);
        }

        @Override
        public JSValue __requireInternal(String path) {
            return require.__requireInternal(path);
        }

        @Override
        public void async(final JSValue action) {
            Async.run("js.async", new Runnable() {
                @Override
                public void run() {
                    action.toFunction().call(null, (Object[]) new JSValue[0]);
                }
            });
        }

        @Override
        public void trigger(final String type, final JSValue data) {
            AJObject argument = JS2AJObject.toAJ(data.toObject());
            AJJavaScriptCoreRuntime.this.trigger(type, argument);
        }

        @Override
        public void exec(final String plugin, final String fn, final JSValue data, final JSFunction callback) {
            AJObject argument = JS2AJObject.toAJ(data.toObject());
            AJJavaScriptCoreRuntime.this.exec(plugin, fn, argument, new Plugin.Callback() {
                @Override
                public void onSuccess(AJObject data) {
                    callback.call(null, false, data);
                }

                @Override
                public void onError(AJObject data) {
                    callback.call(null, true, data);
                }
            });
        }
    }

    private JSContext jsContext;
    private Require require;
    private JSObject jsRuntime;
    private JSFunction dispatchFunction;

    public AJJavaScriptCoreRuntime(final Context context) {
        super(context);

        jsContext = new JSContext();
        jsContext.setExceptionHandler(new JSContext.IJSExceptionHandler() {
            @Override
            public void handle(JSException e) {
                e.printStackTrace();
            }
        });

        require = new Require(context, jsContext);

        JSObject platform = new JSObject(jsContext);
        platform.property("engine", "native");
        platform.property("device", "Android");
        jsContext.property("platform", platform);

        jsContext.evaluateScript("var global = this;");

        NativeGlobals nativeGlobals = new NativeGlobalsImpl(jsContext);
        jsContext.property("native", nativeGlobals);
        jsContext.evaluateScript("global.__trigger = native.trigger;");
        jsContext.evaluateScript("global.async = native.async;");
        jsContext.evaluateScript("global.__exec = native.exec;");
        jsContext.evaluateScript("global.__requireInternal = native.__requireInternal;");

        Timers timers = new TimersImpl(jsContext);
        jsContext.property("native").toObject().property("timers", timers);
        jsContext.evaluateScript("global.setTimeout = native.timers.setTimeout;");
        jsContext.evaluateScript("global.setInterval = native.timers.setInterval;");
        jsContext.evaluateScript("global.clearTimeout = native.timers.clearTimeout;");
        jsContext.evaluateScript("global.clearInterval = native.timers.clearTimeout;");

        Logger logger = new LoggerImpl(jsContext);
        jsContext.property("logger", logger);
        jsContext.evaluateScript("logger.i = function() { logger.__i(Array.prototype.join.call(arguments, ' ')); }");
        jsContext.evaluateScript("logger.w = function() { logger.__w(Array.prototype.join.call(arguments, ' ')); }");
        jsContext.evaluateScript("logger.e = function() { logger.__e(Array.prototype.join.call(arguments, ' ')); }");

        jsContext.property("__httpClient", new HttpClientImpl(context, jsContext));
        jsContext.property("__assetsManager", new AssetsManagerImpl(context, jsContext));
        jsContext.property("__storageManager", new StorageManagerImpl(context, jsContext));
        jsContext.property("__buffersManager", new BuffersManagerImpl(context, jsContext));
        jsContext.property("device", new DeviceImpl(context, jsContext));

        jsContext.evaluateScript("var DEBUG = true;");
        jsContext.evaluateScript("var LOG_LEVEL_INFO = 3;");
        jsContext.evaluateScript("var LOG_LEVEL_WARNING = 2;");
        jsContext.evaluateScript("var LOG_LEVEL_ERROR = 1;");
        jsContext.evaluateScript("var LOG_LEVEL_DISABLED = 0;");
        jsContext.evaluateScript("var LOG_LEVEL = LOG_LEVEL_INFO;");

        try {
            InputStream inputStream = getContext().getAssets().open("js/app.js");
            String source = IOUtils.toString(inputStream);
            IOUtils.closeQuietly(inputStream);

            jsContext.evaluateScript(source);

            JSFunction main = jsContext.property("main").toFunction();
            jsRuntime = main.call().toObject();

            dispatchFunction = jsRuntime.property("run").toFunction();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Semaphore run(final String action, final AJObject data) {
        Log.i("AJ", String.format("Dispatching action %s", action));

        return new Semaphore(new Runnable() {
            @Override
            public void run() {
                dispatchFunction.call(
                        jsRuntime,
                        new JSValue(jsContext, action),
                        JS2AJObject.toJS(jsContext, data)
                );

            }
        });
    }
}

File: platforms/ios/Libs/AJ/AJ/AJJavaScriptCoreRuntime.swift
//
//  AJJavaScriptCoreRuntime.swift
//  AJ
//
//  Created by Bruno Fortunato on 18/02/16.
//  Copyright © 2016 Bruno Fortunato. All rights reserved.
//

import UIKit
import JavaScriptCore
import ApplicaFramework

open class AJJavaScriptCoreRuntime: AJRuntime {

    let jsContext: JSContext
    let timers: AJTimers
    
    var jsRuntime: JSValue?
    
    public override init() {
        jsContext = JSContext()
        timers = AJTimers()
        super.init()
        
        jsContext.exceptionHandler = { context, exception in
            // type of String
            if let exception = exception {
                let stacktrace: String = (exception.objectForKeyedSubscript("stack").toString()) ?? "NA"
                // type of Number
                let lineNumber: JSValue = exception.objectForKeyedSubscript("line") ?? JSValue(object: "", in: context)
                // type of Number
                let column: JSValue = exception.objectForKeyedSubscript("column") ?? JSValue(object: "", in: context)
                let moreInfo = "\n   - in method \(stacktrace)\n   - line number in file: \(lineNumber)\n   - column: \(column)"
                NSLog("JS ERROR \n\(exception) \(moreInfo)")
            } else {
                NSLog("JS ERROR")
            }
        }
        
        let aj_async: @convention(block) (JSValue) -> Void = { action in
            async({
                action.call(withArguments: [])
            })
        }
        
        let aj_setTimeout: @convention(block) (JSValue, Int) -> Int = { action, time in
            return self.timers.setTimeout(action, time)
        }
        
        let aj_setInterval: @convention(block) (JSValue, Int) -> Int = { action, time in
            return self.timers.setInterval(action, time)
        }
        
        let aj_clearTimeout: @convention(block) (Int) -> Void = { timerId in
            return self.timers.clearTimeout(timerId)
        }
        
        let aj_clearInterval: @convention(block) (Int) -> Void = { timerId in
            return self.timers.clearInterval(timerId)
        }
        
        let aj_trigger: @convention(block) (String, JSValue) -> Void = { (store, data) in
            let dict = data.toDictionary() as? [String: AnyObject]
            let state: AJObject = dict != nil ? AJObject(dict: dict!) : AJObject.empty()

            self.tigger(store: store, data: state)
        }
        
        let aj_exec: @convention(block) (String, String, JSValue, JSValue) -> Void = { (plugin, fn, data, callback) in
            let dict = data.toDictionary() as? [String: Any]
            let arguments: AJObject = dict != nil ? AJObject(dict: dict!) : AJObject.empty()
            
            self.exec(plugin: plugin, fn: fn, data: arguments) { (error, result) in
                callback.call(withArguments: [error, result?.toDict() ?? [String: AnyObject]()])
            }
        }
        
        jsContext.globalObject.setObject(jsContext.globalObject, forKeyedSubscript: "global" as (NSCopying & NSObjectProtocol)!)
        
        let platform = ["engine": "native", "device": "iOS"]
        jsContext.globalObject.setObject(platform, forKeyedSubscript: "platform" as (NSCopying & NSObjectProtocol)!)
        
        //used by js components to notify the native parts
        jsContext.globalObject.setObject(unsafeBitCast(aj_trigger, to: AnyObject.self), forKeyedSubscript: "__trigger" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_exec, to: AnyObject.self), forKeyedSubscript: "__exec" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_async, to: AnyObject.self), forKeyedSubscript: "async" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_setTimeout, to: AnyObject.self), forKeyedSubscript: "setTimeout" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_setInterval, to: AnyObject.self), forKeyedSubscript: "setInterval" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_clearTimeout, to: AnyObject.self), forKeyedSubscript: "clearTimeout" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(unsafeBitCast(aj_clearInterval, to: AnyObject.self), forKeyedSubscript: "clearInterval" as (NSCopying & NSObjectProtocol)!)

        jsContext.globalObject.setObject(AJLogger(), forKeyedSubscript: "logger" as (NSCopying & NSObjectProtocol)!)
        jsContext.evaluateScript("logger.i = function() { logger.__i(Array.prototype.join.call(arguments, ' ')); }")
        jsContext.evaluateScript("logger.w = function() { logger.__w(Array.prototype.join.call(arguments, ' ')); }")
        jsContext.evaluateScript("logger.e = function() { logger.__e(Array.prototype.join.call(arguments, ' ')); }")
        
        jsContext.globalObject.setObject(AJHttpClient(runtime: self), forKeyedSubscript: "__httpClient" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(AJAssetsManager(runtime: self), forKeyedSubscript: "__assetsManager" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(AJStorageManager(runtime: self), forKeyedSubscript: "__storageManager" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(AJBuffersManager(), forKeyedSubscript: "__buffersManager" as (NSCopying & NSObjectProtocol)!)
        jsContext.globalObject.setObject(AJDevice(), forKeyedSubscript: "device" as (NSCopying & NSObjectProtocol)!)

        jsContext.evaluateScript("var DEBUG = true;")
        jsContext.evaluateScript("var LOG_LEVEL_INFO = 3;")
        jsContext.evaluateScript("var LOG_LEVEL_WARNING = 2;")
        jsContext.evaluateScript("var LOG_LEVEL_ERROR = 1;")
        jsContext.evaluateScript("var LOG_LEVEL_DISABLED = 0;")
        jsContext.evaluateScript("var LOG_LEVEL = LOG_LEVEL_INFO;")
        
        let appDir = "/assets/js/"
        if let url = Bundle.main.path(forResource: "app", ofType: "js", inDirectory: appDir) {
            let source = try? String(contentsOfFile: url)
            
            if let source = source {
                jsContext.evaluateScript(source)
                let main = jsContext.globalObject.objectForKeyedSubscript("main")
                if let main = main {
                    self.jsRuntime = main.call(withArguments: [])
                } else {
                    fatalError("Main function not found in app.js")
                }
                
                if self.jsRuntime == nil {
                    fatalError("Cannot initialize aj runtime")
                }
                
                NSLog("Runtime initialized: \(String(describing: jsRuntime?.toDictionary()))")
            }
        } else {
            fatalError("app.js not found")
        }
    }
    
    open override func run(action: String, data: AJObject = AJObject.empty()) -> AJSemaphore {
        if let jr = self.jsRuntime {
            let semaphore = AJSemaphore(action: { () -> Void in
                let dict = data.toDict()
                jr.invokeMethod("run", withArguments: [action, dict])
            })
            return semaphore
        }
        
        fatalError("jsRuntime not initialized")
    }
    
}
